streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,644 @@
 
-                needsInitializing.remove(partition);
-                initialized.remove(partition);
-                restorer.setCheckpointOffset(restoreConsumer.position(partition));
+            if (endOffset != null && committedOffset != null) {
+                changelogMetadata.restoreEndOffset = Math.min(endOffset, committedOffset);
 
-                task.reinitializeStateStoresForPartitions(Collections.singleton(partition));
+                log.debug("End offset for changelog {} initialized as {}.", partition, changelogMetadata.restoreEndOffset);
             } else {
-                log.info("Restoring task {}'s state store {} from beginning of the changelog {} ", task.id, restorer.storeName(), partition);
-
-                final long position = restoreConsumer.position(restorer.partition());
-                logRestoreOffsets(restorer.partition(),
-                        position,
-                        restoreToOffsets.get(restorer.partition()));
-                restorer.setStartingOffset(position);
-                restorer.restoreStarted();
+                if (!newPartitionsToRestore.remove(changelogMetadata))
+                    throw new IllegalStateException("New changelogs to restore " + newPartitionsToRestore +
+                        " does not contain the one looking for end offset: " + partition + ", this should not happen.");
+
+                log.info("End offset for changelog {} cannot be found; will retry in the next time.", partition);
             }
         }
 
-        needsRestoring.addAll(initialized);
-    }
+        // try initialize limit offsets for standby tasks for the first time
+        if (!committedOffsets.isEmpty())
+            updateLimitOffsetsForStandbyChangelogs(committedOffsets);
 
-    private void logRestoreOffsets(final TopicPartition partition,
-                                   final long startingOffset,
-                                   final Long endOffset) {
-        log.debug("Restoring partition {} from offset {} to endOffset {}",
-                  partition,
-                  startingOffset,
-                  endOffset);
-    }
+        // add new partitions to the restore consumer and transit them to restoring state
+        addChangelogsToRestoreConsumer(newPartitionsToRestore.stream().map(metadata -> metadata.changelogPartition)
+            .collect(Collectors.toSet()));
 
-    private void refreshChangelogInfo() {
-        try {
-            partitionInfo.putAll(restoreConsumer.listTopics());
-        } catch (final TimeoutException e) {
-            log.debug("Could not fetch topic metadata within the timeout, will retry in the next run loop");
+        // if it is in the active restoring mode, we immediately pause those standby changelogs
+        // here we just blindly pause all (including the existing and newly added)
+        if (state == ChangelogReaderState.ACTIVE_RESTORING) {
+            pauseChangelogsFromRestoreConsumer(standbyRestoringChangelogs());
```

Comment:
Looking at the current call trace from rebalance callback, we will create new standby / active tasks at the end of a rebalance and then in state manager when we register the stores (see `AssignedTasks#initializeNewTasks -> Task#initializeStateStores`. So even when we are in `ACTIVE_RESTORING` state we can still register new standby changelogs.

Reply:
I see. Can we just immediately pause new standby changelogs as they are added, since we will also go through the `ACTIVE_RESTORING` state after a rebalance IIUC. This keeps that logic encapsulated so we don't have to be considering when/where standbys might be added, or thinking about the standbys at all, while initializing the active changelogs